Rewrite Fluid Mixins

ThistleSifter opened the pull request:
Here is the rewrite of the Fluid Mixin library. Most of the important information is in the documentation which I rewrote with more examples. Here's a short summary of the changes
- Mixins are now in their own namespace and the `finale` namespace has been left *almost* untouched*
- Creating an object in the original `finale` namespace does not result in mixins or the fluid interface becoming available**
- New mixin class structure with `FCM` and `FCX` mixin classes.
- Added helper functions to add more control when defining mixins. Using them can add some complexity, but I think it's worth it because it makes using the mixins more user-friendly.
- Rewrote documentation with improved examples, including how to have private properties without a memory leak.

*I added a single property to keep track of which `FC` classes had their metatables modified.
** The only scenario in which an `FC` object could be unintentionally converted to an `FCM` object is if the `FC` object is passed to an `FCM` method which then returns that same object, as mixins are activated on the return of an `FC` object from an `FCM` method. But this is quite a small catch and I doubt it will cause any problems.

Also note that when accessing a method statically (eg `finalemix.FCMControl.SetText`), the proxied method will be returned, not the original (proxying takes care of automatically promoting all returned `FC` objects to `FCM` and handles the fluid interface). This was mainly to ensure consistency of return behaviour across all mixin methods, regardless of context.

rpatters1 commented on src/library/mixin.lua:
```diff
@@ -4,227 +4,368 @@
-**Adding New Methods and Properties**
-Including the mixin library instantly enables the addition of new methods and properties. The example below demonstrates adding a new property and a new method which accesses that property.
+-- Customised mixin that extends FCXString. Still has the same constructor signature as FCString
+local fcxstr2 = finale.FCXMyCustomString()
```
Is `finale` correct here? Did you mean `finalemix` or `fcsstr` instead?

ThistleSifter replied:
Good catch, fixed.

rpatters1 reviewed (commented):
This is looking very good to me. I will find some time to test it in the coming week.

Nick-Mazuk commented:
This is looking awesome to me. I really like the clear separation between the `finale` and `finalemix` namespaces.

Nothing stands out to me that needs fixing. Though I'm not sure if I'm going to get time this week to test it. @rpatters1 , if you're able to test before me, feel free to go ahead and merge as well.

I'm not sure if the docs parser has been tested with using the ``` notation to create codeblocks, but it should work. If it doesn't, I'll update that parser to make sure it works.

ThistleSifter commented:
I was working on updating the control mixins PR and I realised that some functions would be better placed in the other libraries so I moved `is_finale_object` to `general_library` and I added an `unpack` function to `utils`.

ThistleSifter commented:
I was doing some thinking about state management, or more specifically state persistence, and I wanted to discuss whether the mixin library should do anything about it.

Here's an example:
```lua
local dialog = finale.FCCustomLuaWindow()
local edit = dialog:CreateEdit(10, 10)

dialog:RegisterHandleCommand(function(ctrl)
    print(ctrl:GetControlID() == edit:GetControlID()) --true
    print(ctrl == edit) --false
end)

dialog:ExecuteModal(nil)
```

If the first two lines were replaced with the following, it starts to get a little more troublesome as all mixin data would be lost in the handler:
```lua
local dialog = finalemix.FCXMyDialog()
local edit = dialog:CreateCustomEdit()
```

For controls and dialogs, it's quite an easy situation to fix, since I think it would be expected that the control passed to the handler would be the same as the one created (including if it's an `FCX` mixin). In this case, I'd just write some mixin method intercepts to store the original controls, compare IDs, and make sure the original is always returned when needed.

But what about something like this (which is a little contrived, but in a more complex plugin, it would be possible to run into a similar scenario while traversing function calls or iterators):
```lua
local m = finalemix.FCMMeasure()
local mm = finalemix.FCMMeasure()

-- Load the same measure
m:Load(1)
mm:Load(1)

m.Foo = 1
mm.Foo = 2 -- m.Foo is still 1
```

The PDK is quite large and I'm not familiar with a lot of it, so I don't know if anyone would expect any mixin state to be preserved between the two. My thinking is probably not, since it's not expected that state would persist between `FC` objects either, even if they do share some sort of identifier.

What do you think? I'm basically trying to work out if this needs to be handled by the library, or whether the dialog controls scenario is an outlier which can be treated separately and handled by itself.

rpatters1 commented:
If I understand your question, you are asking if there could be a situation where all instances of a particular class need to share a property value. (Or all instances loaded from a particular Finale data structure.) I can't see that being needed in general, but perhaps specific classes might use it. Whether to implement it in the general case would depend a great deal on how complex the code is (remembering that every line of code is a liability) and if it carries any appreciable performance cost.

I also encourage you to start thinking in terms of (and testing with) the expanded capabilities of RGP Lua (if you haven't already) since that is the future. RGP Lua allows a script to maintain state between invocations from Finale. So state will become increasingly important.

One of the primary classes I want to implement (or see implemented) as an `FCM` or `FCX` class is a wrapper class for `FCCustomLuaWindow`. One of the primary uses of the wrapper class will be to maintain state of the window so that every time the user opens it, it returns to the same position it was in when the user closed it and is in the same state as when the user closed it. (I see this as primarily useful for modeless windows, but it could be useful for modal windows as well.) This means, at a minimum, storing the x,y position of the window plus allowing the user to provide additional context-specific state.

You can see *ad hoc* examples of this in [`transpose_by_step.lua`](https://github.com/finale-lua/lua-scripts/blob/master/src/transpose_by_step.lua) and [`transpose_chromatic.lua`](https://github.com/finale-lua/lua-scripts/blob/master/src/transpose_chromatic.lua). (A much more complex example is the [class browser](https://github.com/finale-lua/rgpluaclassbrowser/blob/master/rgpluaclassbrowser.lua).

I would like to formalize this, if not in the mixin code per se, in a custom mixin wrapper for `FCCustomLuaWindow`. I'm not sure how directly this comment relates to your question about state, but it seems like it does because of maintaining the controls in the state. This might allow the window to be automatically rehydrated from the state without the need of user code, a definite nice-to-have.

rpatters1 commented:
Should this code work, or is an additional `FCXCustomLuaWindow.lua` module required?

```lua
local mixin = require("library.mixin")
local dialog1 = mixin.FCXCustomLuaWindow()
print("ClassName of dialog1 is "..dialog1:ClassName())
local dialog2 = mixin.FCXCustomLuaWindow()
print("ClassName of dialog2 is "..dialog2:ClassName())
```

Right now this code produces an error message:

```lua
"ERROR: ...Projects (XCode, Lazarus)/Lua Scripts/src/mixin_test.lua:27: attempt to call field 'FCXCustomLuaWindow' (a nil value)"
```
If an additional module is required, maybe that needs to be spelled out in the documentation. Also line `34` seems like it should be a triple-backtick rather than a single-backtick.

Otherwise, I'm cool with merging this, unless @ThistleSifter  wants to work more magic on it.

ThistleSifter commented:
Thanks for the detailed explanation. I think I'll leave the library as is for now and handle any specific cases as needed. There's more than enough magic as is!

The example you posted above is meant to fail in the way that it did. Basically, creating an `FCM` object gives you what the library previously did: an editable object with a fluid interface. But then it also loads any defined `FCM` mixins, and these are optional. When an `FCX` object is created, it first creates the underlying `FCM` and then applies modifications in layers. If the requested `FCX` mixin doesn't exist, it will return `nil`. If any of its parents don't exist, an error will be thrown during loading.

For example, take the following hierarchy:
`FCXMixC` > `FCXMixB` > `FCXMix` > `FCMString`

When calling `finalemix.FCXMixC()`, it would first create the `FCMString` object, then it would apply the `FCXMixA` mixin, then apply `FCXMixB` on top of that, and then apply `FCXMixC` on top of that.

Essentially, anything that is defined in an `FCX` mixin can just be applied manually, like so (new properties, `FCX` mixins, and manual modifications all share the same storage):
```lua
local str = finalemix.FCMString()

str.SetLuaString = function(self, str) end
str.Foo = 'foo'
str.NewMethod = function(self) end
```

Enabling them to be formally defined as `FCX` mixins (and being able to access the current 'class' in `MixinClassName`) just adds some structure to what could otherwise be the wild west.


I have updated the documentation to hopefully clear some of these things up.


Also, following on from what you were saying about state management, it's entirely conceivable that a plugin could be completely encapsulated in a customised dialog object. This could also open up interesting possibilities with interactions between plugins.
